src/qibo/models/error_mitigation.py

```diff
@@ -432,6 +431,8 @@ def CDR(
         len(signature(model).parameters) - 1
     )  # first arg is the input and the *params afterwards
     params = backend.cast(local_state.random(nparams), backend.precision)
+    if backend.name == "pytorch":
+        params.requires_grad = True
```

Comment:
All this specific case handling is not ideal though, can't we have `PyTorchBackend.cast` enabling gradients by default? I mean, in general, gradients are the main reason why you would use the backend in the first place.

Reply:
with @renatomello we decided to set require_gradients=False by default (we changed it from the first version) to avoid putting gradients everywhere like in the circuit initial state.

Reply:
I see, it's just that `pytorch` cast is now asymmetric compared to other backends, which sometimes forces you to deal with this case handling. One thing, I was wondering is why don't we use [`torch.is_grad_enabled`](https://pytorch.org/docs/stable/generated/torch.is_grad_enabled.html):
``` python
def cast(self, ...):
    requires_grad = self.np.is_grad_enabled()
    ....
```

Reply:
but still you will activate gradients for the initial state of the circuit and many other variables that do not require them

Reply:
I don't know, in general I am not understanding why it is a problem to have an `initial_state` tensor with `requires_grad=True`, which other cases are you referring to in particular can you provide some examples?

Reply:
Calculating gradients of immutable objects is both a waste of resources and has no physical meaning.

```In [1]: import numpy as np
   ...: import torch
   ...: 
   ...: from qibo import Circuit, gates, set_backend
   ...: from qibo.backends import PyTorchBackend
   ...: from qibo.quantum_info import infidelity
   ...: 
   ...: set_backend("pytorch")
   ...: backend = PyTorchBackend()
   ...: 
   ...: 
   ...: nqubits, nepochs = 2, 1
   ...: dims = 2**nqubits
   ...: 
   ...: target_state = np.random.rand(dims)
   ...: target_state /= np.linalg.norm(target_state)
   ...: target_state = backend.cast(target_state, requires_grad=True)
   ...: 
   ...: initial_state = np.random.rand(dims)
   ...: initial_state /= np.linalg.norm(initial_state)
   ...: initial_state = backend.cast(initial_state, requires_grad=True)
   ...: 
   ...: params = backend.cast(np.random.rand(nqubits), dtype=float, requires_grad=True)
   ...: 
   ...: optimizer = torch.optim.Adam([params])
   ...: 
   ...: circuit = Circuit(nqubits)
   ...: circuit.add(gates.RX(q, 0.0) for q in range(nqubits))
   ...: 
   ...: for _ in range(nepochs):
   ...:     optimizer.zero_grad()
   ...:     circuit.set_parameters(params)
   ...:     state = circuit(initial_state=initial_state).state()
   ...:     loss = infidelity(state, target_state)
   ...:     loss.backward(retain_graph=True)
   ...:     print(initial_state.grad)
   ...:     print(target_state.grad)
[Qibo 0.2.13|INFO|2024-10-02 10:29:36]: Using pytorch backend on cuda:0
tensor([-0.7136-0.1033j, -0.0798-0.2461j, -0.9491-0.0770j, -1.3825+0.0783j],
       dtype=torch.complex128)
tensor([-0.0810+0.1756j, -0.1235+0.2421j, -1.0581-0.0375j, -1.4742-0.0753j],
       dtype=torch.complex128)
```

Reply:
Ok so, for the user it's just a matter of preparing the initial state (which is anyway not a widely used feature in my personal experience) as:
``` python
initial_state = torch.randn(2 ** nqubits, requires_grad=False)
```

In any case, regardless of the philosophical choice between `requires_grad=True` or `False` by default, I believe we should think twice about creating an asymmetry with a widely used backend method such as `backend.cast`

Reply:
Yes, that's precisely the asymmetry, you add an argument to the `cast` method that is not present in any other. This means that if you implement a function in `qibo` that needs to cast an array, it will be painless only if the default value is the one you need, otherwise you will have to check if the backend is pytorch, as it happens in this case:
``` python
if backend.name == "pytorch":
    backend.cast(param, requires_grad=True)
else:
    backend.cast(param)
```
Therefore, either we separate the `cast_parameter` and `cast` in all backends, which in most cases, if not all, are just going to coincide, or you rely on some internal variable like the `torch.is_grad_enabled`.

Reply:
Correct me if I'm wrong, but my understanding is that `torch.is_grad_enabled` is about gradients being enabled globally. 
We can't rely on that because what determines if a gate should have grads is its `trainable` attribute (this is important e.g. in seesaw algorithms), and what determines if a specific tensor requires grads is its `requires_grad` attribute.
The solution you're seeing was specifically introduced to remove the enabling of grads globally because this was creating a bunch of issues like the ones in the example I showed.
Creating a `cast_parameter` for every backend does not necessarily solve the issues you're raising because it doesn't cover cases like when a gate like `gates.Unitary` requires grad, or when a statevector / density matrix requires grad.

One way that I could see to cover all issues is the following:

- Remove `requires_grad` from `PyTorchBackend.cast`
- Create `AbstractBackend.cast_parameter`
- Overwrite `cast_parameter` in `PyTorchBackend` (technically already done)
- Explicitly mention in the documentation that if one wants to optimize something that is not a parameterized gate, it should do something like the following:
```python3
backend = PyTorchBackend()

matrix = backend.cast(matrix)
matrix.requires_grad_()
```

Reply:
Yes, it is like that, `torch.is_grad_enabled=False` if you are inside a `with torch.no_grad()` context, but I get it that it's a little bit clunky. Ok got it, I am ok with not having gradients enabled by default, but I would still get rid of the `requires_grad` argument of the `cast`, which is set to true just a handful of times
```
doc/source/code-examples/advancedexamples.rst:        torch.rand(2, dtype=torch.float64), requires_grad=True
src/qibo/backends/pytorch.py:            return self.np.tensor(x, dtype=self.parameter_dtype, requires_grad=True)
tests/test_backends_torch_gradients.py:    params = backend.np.rand(2, dtype=backend.np.float64, requires_grad=True)
tests/test_backends_torch_gradients.py:    param = backend.np.tensor([0.1], dtype=backend.np.float64, requires_grad=True)
tests/test_models_variational.py:            initial_parameters, dtype=backend.np.float32, requires_grad=True
tests/test_models_variational.py:            initial_p, dtype=backend.np.float32, requires_grad=True
src/qibo/models/error_mitigation.py:        params.requires_grad = True
src/qibo/models/error_mitigation.py:        params.requires_grad = True
```
and we can easily work around that by moving inside the backend.
@Simone-Bordoni can you please restore this as it was before, namely setting the `requires_grad=True` inside the `_curve_fit` function? Sorry for the confusion. 
I am planning to transform that `_curve_fit` into a backend method in a future PR anyway, such that we don't have to deal with checking the backend used.

Reply:
@Simone-Bordoni can you restore this as it was in the beginning, i.e. casting to parameter for `pytorch`inside the `_curve_fit` funcion?